libazureinit/src/provision/user.rs

```diff
@@ -5,7 +5,57 @@ use std::process::Command;
 
 use tracing::instrument;
 
-use crate::error::Error;
+use crate::{error::Error, imds::PublicKeys};
+
+/// Configuration for the provisioned user.
+#[derive(Clone)]
+pub struct User {
+    pub(crate) name: String,
+    pub(crate) groups: Vec<String>,
+    pub(crate) ssh_keys: Vec<PublicKeys>,
+    pub(crate) password: Option<String>,
+}
+
+impl core::fmt::Debug for User {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // This is manually implemented to avoid printing the password if it's set
+        f.debug_struct("User")
+            .field("name", &self.name)
+            .field("groups", &self.groups)
+            .field("ssh_keys", &self.ssh_keys)
+            .field("password", &self.password.is_some())
+            .finish()
+    }
+}
+
+impl User {
+    /// Configure the user being provisioned on the host.
+    pub fn new(
+        name: impl Into<String>,
+        ssh_keys: impl Into<Vec<PublicKeys>>,
+    ) -> Self {
+        Self {
+            name: name.into(),
+            groups: vec!["wheel".into()],
+            ssh_keys: ssh_keys.into(),
+            password: None,
+        }
+    }
+
+    /// Set a password for the user; this is optional.
+    pub fn with_password(mut self, password: impl Into<String>) -> Self {
+        self.password = Some(password.into());
+        self
+    }
+
+    /// A list of supplemental group names to add the user to.
+    ///
+    /// If any of the groups do not exist on the host, provisioning will fail.
```

Comment:
We could, instead, probe for groups that exist and only add the user to those groups. I don't love that because it feels spooky and unexpected ("I asked it to add the user to these groups and it didn't, but also didn't fail"), but on the other hand you wouldn't have to hand-tune the list for each distribution.